source/mir/cv/imgproc/imfilter.d

```diff
@@ -0,0 +1,459 @@
+pragma(inline, false)
+int separable_imfilter
+(
+    float* inputbuf,
+    float* hmask,
+    float* vmask,
+    float* outputbuf,
+    size_t rows,
+    size_t cols,
+    size_t hsize,
+    size_t vsize,
+    int border_handling = BORDER_REPLICATE
+)
+{
+    return separable_imfilter_impl
+        (inputbuf, hmask, vmask, outputbuf, rows, cols, hsize, vsize, border_handling);
+}
+
+package(mir.cv):
+
+template Horizontal_kernel_func(alias InstructionSet)
+{
+    alias Horizontal_kernel_func = void function(InstructionSet.Scalar*, // input pointer.
+        InstructionSet.Vector*, // kernel mask pointer.
+        InstructionSet.Scalar*, // output pointer.
+        size_t); // size of the kernel mask.
+}
+
+template Vertical_kernel_func(alias InstructionSet)
+{
+    alias Vertical_kernel_func = void function(InstructionSet.Scalar*, // input pointer.
+        InstructionSet.Vector*, // kernel mask pointer.
+        InstructionSet.Scalar*, // output pointer.
+        size_t, // size of the kernel mask.
+        size_t); // rowstride to read another row element.
+}
+
+pragma(inline, true)
+auto max(T)(T t1, T t2) {
+    return (t1 > t2) ? t1 : t2;
+}
+
+int separable_imfilter_impl(T)
+(
+    T* inputbuf,
+    T* hmask,
+    T* vmask,
+    T* outputbuf,
+    size_t rows,
+    size_t cols,
+    size_t hsize,
+    size_t vsize,
+    int border_handling = BORDER_REPLICATE
+)
+{
+    auto tempbuf = cast(T[])alignedAllocate(rows * cols * T.sizeof, 16); // temporary buffer, used to store horizontal filtering
+
+    scope (exit)
+    {
+        deallocate(tempbuf);
+    }
+
+    auto input = inputbuf.sliced(rows, cols).universal;
+    auto temp = tempbuf.sliced(rows, cols).universal;
+    auto output = outputbuf.sliced(rows, cols).universal;
+    auto hm = hmask.sliced(hsize).universal;
+    auto vm = vmask.sliced(vsize).universal;
+
+    if (avx) // add version (compile time flag if avx should be supported)
+    {
+        inner_filtering_impl!(AVX!T)(input, temp, hm, vm, output,
+            &apply_horizontal_kernel_simd!(AVX!T), &apply_vertical_kernel_simd!(AVX!T));
+    }
+    else version (sse)
+    {
+        inner_filtering_impl!(SSE!T)(input, temp, hm, vm, output,
+            &apply_horizontal_kernel_simd!(SSE!T), &apply_vertical_kernel_simd!(SSE!T));
+    }
+    else
+    {
+        inner_filtering_impl!(Non_SIMD!T)(input, temp, hm, vm, output,
+            get_horizontal_kernel_for_mask!T(hsize), get_vertical_kernel_for_mask!T(vsize));
+    }
+
+    if (border_handling != BORDER_REPLICATE)
+    {
+        return ERROR_INVALID_BORDER_METHOD;
+    }
+
+    borders_replicate_impl(output, hsize, vsize);
+
+    return 0;
+}
+
+pragma(inline, true) void inner_filtering_impl(alias InstructionSet)(
+    Slice!(Universal, [2], InstructionSet.Scalar*) input, Slice!(Universal,
+    [2], InstructionSet.Scalar*) temp, Slice!(Universal, [1],
+    InstructionSet.Scalar*) hmask, Slice!(Universal, [1],
+    InstructionSet.Scalar*) vmask, Slice!(Universal, [2],
```

Comment:
Can we have `Contiguous` input and use `universal` internally because this function assumes that slices has the same strides